### ETL Process (Extract, Transform & Load) - Stream Video Games Platform
#### Data: User_Items

We will start by understanding, transforming, and making the data available by identifying issues, applying cleaning techniques, and preprocessing. Finally, we will store the transformed data for future explorations.

We will understand the context and the necessary information for our analysis, ensuring data consistency.

#### Importing necessary libraries:

- `numpy`: For working with arrays and matrices.
- `pandas`: For tabular data analysis.
- `os`: Allows interaction with the operating system.
- `gdown`: Facilitates downloading files from Google Drive.
- `json`: Commonly used for data exchange.
- `warnings`: Filters and suppresses warnings.
- `ast`: Useful for parsing Python code at a deeper level than string parsing.


In [1]:
import numpy as np  # For array and matrix operations
import pandas as pd  # For tabular data analysis
from pandas import json_normalize  # For normalizing JSON data

import os  # For interacting with the operating system

import gdown  # For downloading files from Google Drive

import json  # Common data exchange format

import warnings
warnings.filterwarnings("ignore")  # Suppresses warnings

import ast  # For parsing Python code



### **1. Load Dataset**

Data Source: australian_user_items.json

We'll download the file locally since it needs to be loaded in a specific way due to its structure.

This file is available on Google Drive with shared access. You can download it from the following link:


Dataset: https://drive.google.com/uc?export=download&id=1BzbeJv1878pa2lQMspr5y03dUdIXaYyX


In [26]:
path= '../Data_Files/raw_Data/users_items.json'


with open(path, 'r', encoding='utf-8') as f:
    data = f.readlines()

# vamos a convertir las lineas en registros
records = [eval(line.strip()) for line in data]

# Creamos el DataFrame a partir de los records o registros
df_UserItems= pd.DataFrame(records)

In [27]:
respaldo = df_UserItems

### 2nd Method for Uploading Raw Data: Gzip Format

In [7]:
# import gzip 
# import ast
# import pandas as pd
# data = []
# try:
#     with gzip.open('users_items.json.gz', 'rb') as d:
#         for line in d:
#             data.append (ast.literal_eval(line.decode('utf-8')))
#     Users_Items = pd.DataFrame (data)
# finally:
#     d.close()

## Explore and understand the dataset


In [29]:
# Quick view of the table
print('DataFrame - First & Last Rows:')
df_UserItems


DataFrame - First & Last Rows:


user_id  items_count           steam_id  \
0            76561197970982479          277  76561197970982479   
1                      js41637          888  76561198035864385   
2                    evcentric          137  76561198007712555   
3                   Riot-Punch          328  76561197963445855   
4                        doctr          541  76561198002099482   
...                        ...          ...                ...   
88305        76561198323066619           22  76561198323066619   
88306        76561198326700687          177  76561198326700687   
88307  XxLaughingJackClown77xX            0  76561198328759259   
88308        76561198329548331            7  76561198329548331   
88309        edward_tremethick            0  76561198331598578   

                                                user_url  \
0      http://steamcommunity.com/profiles/76561197970...   
1                   http://steamcommunity.com/id/js41637   
2                 http://steamcommunity.com/id/evcentric   
3                http://steamcommunity.com/id/Riot-Punch   
4                     http://steamcommunity.com/id/doctr   
...                                                  ...   
88305  http://steamcommunity.com/profiles/76561198323...   
88306  http://steamcommunity.com/profiles/76561198326...   
88307  http://steamcommunity.com/id/XxLaughingJackClo...   
88308  http://steamcommunity.com/profiles/76561198329...   
88309     http://steamcommunity.com/id/edward_tremethick   

                                                   items  
0      [{'item_id': '10', 'item_name': 'Counter-Strik...  
1      [{'item_id': '10', 'item_name': 'Counter-Strik...  
2      [{'item_id': '1200', 'item_name': 'Red Orchest...  
3      [{'item_id': '10', 'item_name': 'Counter-Strik...  
4      [{'item_id': '300', 'item_name': 'Day of Defea...  
...                                                  ...  
88305  [{'item_id': '413850', 'item_name': 'CS:GO Pla...  
88306  [{'item_id': '11020', 'item_name': 'TrackMania...  
88307                                                 []  
88308  [{'item_id': '304930', 'item_name': 'Unturned'...  
88309                                                 []  

[88310 rows x 5 columns]

## **Primary Analysis**
1. No missing values are observed in any column of the 'User_items' dataset.

2. The data collection process seems optimized as there are no missing values.

3. The 'items' column contains nested lists.

4. Further in-depth analysis will be performed.


In [30]:
# Check how many rows have null values in all columns
print(f'Total rows in df_UserItems: {df_UserItems.shape[0]}')
print(f'Total rows in df_UserItems with null values in all columns: {df_UserItems.isnull().all(axis=1).sum()}')


Total rows in df_UserItems: 88310
Total rows in df_UserItems with null values in all columns: 0


In [31]:
# General information about the DataFrame
print("General DataFrame Information:\n")
df_UserItems.info()


General DataFrame Information:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88310 entries, 0 to 88309
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      88310 non-null  object
 1   items_count  88310 non-null  int64 
 2   steam_id     88310 non-null  object
 3   user_url     88310 non-null  object
 4   items        88310 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.4+ MB


### **# Insights**
1. The **'items'** column contains a list with nested data (list of dictionaries).

2. We will normalize the column, resulting in a column for each key in the dictionary, maintaining traceability of the columns **('steam_id', 'items_count', 'user_id',** and **'user_url').**


In [32]:
# Normalize the 'items' column
df_UserItems = pd.json_normalize(records, record_path=['items'], meta=['steam_id', 'items_count', 'user_id', 'user_url'])


In [33]:
df_UserItems

item_id                      item_name  playtime_forever  \
0            10                 Counter-Strike                 6   
1            20          Team Fortress Classic                 0   
2            30                  Day of Defeat                 7   
3            40             Deathmatch Classic                 0   
4            50      Half-Life: Opposing Force                 0   
...         ...                            ...               ...   
5153204  346330                   BrainBread 2                 0   
5153205  373330                    All Is Dust                 0   
5153206  388490  One Way To Die: Steam Edition                 3   
5153207  521570          You Have 10 Seconds 2                 4   
5153208  519140                     Minds Eyes                 3   

         playtime_2weeks           steam_id items_count            user_id  \
0                      0  76561197970982479         277  76561197970982479   
1                      0  76561197970982479         277  76561197970982479   
2                      0  76561197970982479         277  76561197970982479   
3                      0  76561197970982479         277  76561197970982479   
4                      0  76561197970982479         277  76561197970982479   
...                  ...                ...         ...                ...   
5153204                0  76561198329548331           7  76561198329548331   
5153205                0  76561198329548331           7  76561198329548331   
5153206                3  76561198329548331           7  76561198329548331   
5153207                4  76561198329548331           7  76561198329548331   
5153208                3  76561198329548331           7  76561198329548331   

                                                  user_url  
0        http://steamcommunity.com/profiles/76561197970...  
1        http://steamcommunity.com/profiles/76561197970...  
2        http://steamcommunity.com/profiles/76561197970...  
3        http://steamcommunity.com/profiles/76561197970...  
4        http://steamcommunity.com/profiles/76561197970...  
...                                                    ...  
5153204  http://steamcommunity.com/profiles/76561198329...  
5153205  http://steamcommunity.com/profiles/76561198329...  
5153206  http://steamcommunity.com/profiles/76561198329...  
5153207  http://steamcommunity.com/profiles/76561198329...  
5153208  http://steamcommunity.com/profiles/76561198329...  

[5153209 rows x 8 columns]

In [34]:
# Find duplicate records
duplicates = df_UserItems[df_UserItems.duplicated(subset=['steam_id', 'item_id', 'playtime_forever'])]
duplicates.shape


(59117, 8)

In [35]:
# Remove duplicates
df_UserItems = df_UserItems.drop_duplicates(keep='first')


In [36]:
# Convert 'playtime_forever' column from minutes to hours
df_UserItems['hours_game'] = df_UserItems['playtime_forever'] / 60

# Round the values to two decimal places (optional)
df_UserItems['hours_game'] = df_UserItems['hours_game'].round(2)


In [37]:
# Get the top 5 maximum values
top_5_max = df_UserItems.nlargest(5, 'hours_game')

# Get the top 5 minimum values
top_5_min = df_UserItems.nsmallest(5, 'hours_game')

# Display the results
print("Top 5 maximum values:")
print(top_5_max[['hours_game']])

print("\nTop 5 minimum values:")
print(top_5_min[['hours_game']])


Top 5 maximum values:
         hours_game
587715     10712.88
2499068    10588.25
4075729    10540.87
1495340    10223.52
1836985    10001.13

Top 5 minimum values:
   hours_game
1         0.0
3         0.0
4         0.0
5         0.0
6         0.0


In [38]:
# Count the rows with a value of 0 in the 'hours_game' column
zero_rows = (df_UserItems['hours_game'] == 0).sum()

# Print the result
print(f"There are {zero_rows} rows with a value of 0 in the 'hours_game' column.")


There are 1847730 rows with a value of 0 in the 'hours_game' column.


In [39]:
# Remove rows with a value of 0 in the 'hours_game' column
df_UserItems = df_UserItems[df_UserItems['hours_game'] != 0]

# Remove irrelevant columns
df_UserItems = df_UserItems.drop(['playtime_2weeks','user_url','items_count','item_name','steam_id','playtime_forever'], axis=1)
df_UserItems.columns


Index(['item_id', 'user_id', 'hours_game'], dtype='object')

## **# Insights:**
1. Remove rows with null values in this column for query optimization reasons.

2. There are 1,847,730 records with a value of 0 in the 'hours_game' column.

3. Irrelevant columns are also removed for query optimization reasons.


In [40]:
df_UserItems

item_id            user_id  hours_game
0            10  76561197970982479        0.10
2            30  76561197970982479        0.12
8           300  76561197970982479       78.88
9           240  76561197970982479       30.88
10         3830  76561197970982479        5.55
...         ...                ...         ...
5153202  304930  76561198329548331       11.28
5153203  227940  76561198329548331        0.72
5153206  388490  76561198329548331        0.05
5153207  521570  76561198329548331        0.07
5153208  519140  76561198329548331        0.05

[3246375 rows x 3 columns]

In [41]:
df_UserItems.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3246375 entries, 0 to 5153208
Data columns (total 3 columns):
 #   Column      Dtype  
---  ------      -----  
 0   item_id     object 
 1   user_id     object 
 2   hours_game  float64
dtypes: float64(1), object(2)
memory usage: 99.1+ MB


## Clean data files are saved in the following formats for availability: CSV/Json/Parquet


In [42]:
# The files are stored locally
# df_UserItems.to_csv('user_items_clean.csv', index=False)
# df_UserItems.to_json('user_items_clean.json', orient='records', lines=True)
df_UserItems.to_parquet('../Data_Files/Cleaned_Data/2_user_items_clean.parquet', index=False)
